# 1. Import libraries and data

In [1]:
# Import Libraries 

import pandas as pd
import numpy as np
import os

In [2]:
# Define Path

path = r'C:\Users\dolson\OneDrive - BEACON TECHNOLOGIES\Desktop\Instacart Basket Analysis 10.2022'

In [3]:
# Import Customers CVS Data

df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))


In [4]:
# look as Customers datafile
df_customers

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

In [5]:
df_customers.shape

(206209, 10)

# 2. Wrangle Data 

In [6]:
# Rename columns
df_customers.rename(columns = {'First Name':'first_name','Surnam':'last_name','Gender':'gender', 'STATE':'state', 
                            'Age': 'age','n_dependants': 'no_of_dependants','fam_status': 'marital_status'}, inplace = True)

In [7]:
# check new column names
df_customers

user_id first_name last_name  gender           state  age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        no_of_dependants marital_status  income  
0                      3        married  165665  
1                      0         single   59285  
2                      2        married   99568  
3                      0         single   42049  
4                      1        married   40374  
...                  ...            ...     ...  
206204                 1        married  148828  
206205                 3        married  168639  
206206                 2        married   53374  
206207                 1        married   99799  
206208                 1        married   57095  

[206209 rows x 10 columns]

# 3. Data Quality Consistency checks

In [8]:
# check data consistency on customers dataframe
df_customers.describe()

user_id            age  no_of_dependants         income
count  206209.000000  206209.000000     206209.000000  206209.000000
mean   103105.000000      49.501646          1.499823   94632.852548
std     59527.555167      18.480962          1.118433   42473.786988
min         1.000000      18.000000          0.000000   25903.000000
25%     51553.000000      33.000000          0.000000   59874.000000
50%    103105.000000      49.000000          1.000000   93547.000000
75%    154657.000000      66.000000          3.000000  124244.000000
max    206209.000000      81.000000          3.000000  593901.000000

In [9]:
#check for duplicates
df_customers[df_customers.duplicated()]

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, no_of_dependants, marital_status, income]
Index: []

# no duplicates

In [10]:
# Check data types 
df_customers.dtypes

user_id              int64
first_name          object
last_name           object
gender              object
state               object
age                  int64
date_joined         object
no_of_dependants     int64
marital_status      object
income               int64
dtype: object

In [11]:
# Check for missing values
df_customers.isnull().sum()

user_id                 0
first_name          11259
last_name               0
gender                  0
state                   0
age                     0
date_joined             0
no_of_dependants        0
marital_status          0
income                  0
dtype: int64

In [12]:
# view missing data 
df_nan = df_customers[df_customers['first_name'].isnull()==True]

In [13]:
df_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no_of_dependants marital_status  income  
53                     2        married   41709  
73                     0         single   82518  
82                     3        married  117099  
99                     1        married  155673  
105                    3        married  151819  
...                  ...            ...     ...  
206038                 3        married   87783  
206044                 2        married  108488  
206090                 1        married   45275  
206162                 0         single   56325  
206171                 2        married   59222  

[11259 rows x 10 columns]

# First_name has 11259 Nan (missing values) will leave and use user_id to identify customers

In [14]:
# Check for mixed data types

for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


# first_name has mixed data type

In [15]:
# check for the data type of 'first_name'
df_customers['first_name'].dtype

dtype('O')

In [16]:
# Convert first_name to string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [17]:
# Check for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

# no other mixed data types

# Memory issues reduce datatypes on columns in customers and  dataframe

In [18]:
#df_customers['user_id'] = df_customers['user_id'].astype('str')

In [19]:
df_customers['age'] = df_customers['age'].astype('int8')

In [20]:
df_customers['no_of_dependants'] = df_customers['no_of_dependants'].astype('int8')

In [21]:
df_customers['income'] = df_customers['income'].astype('int32')

In [22]:
df_customers.dtypes

user_id              int64
first_name          object
last_name           object
gender              object
state               object
age                   int8
date_joined         object
no_of_dependants      int8
marital_status      object
income               int32
dtype: object

# Merge data with the rest of prepared instacart data  

In [23]:
# Import orders_products_merged_flags pkl

df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_flags.pkl'))

In [24]:
# Check datafile 
#df_ords_prods_merge.head()

In [25]:
# Check the shape
df_ords_prods_merge.shape

(32404859, 23)

In [26]:
# Check datatype of common column user_id 
df_ords_prods_merge['user_id'].dtype

dtype('int64')

In [27]:
# Check datatypes of df_ords_prods_merge 
df_ords_prods_merge.dtypes

order_id                              int64
user_id                               int64
order_number                          int64
order_day_of_week                     int64
order_hour_of_day                     int64
days_since_prior_order              float64
new_customer                           bool
product_id                          float64
add_to_cart_order                   float64
reordered                           float64
product_name                         object
aisle_id                              int64
department_id                         int64
prices                              float64
price_range_loc                      object
busiest_days                         object
busiest_period_of_day                object
max_order                             int64
loyalty_flag                         object
mean_price                          float64
spending_flag                        object
median_days_since_previous_order    float64
Frequency_flag                  

In [28]:
# memory issues reduce datatypes in df_ords_prods_merge dataframe 

In [30]:
df_ords_prods_merge['order_number'] = df_ords_prods_merge['order_number'].astype('int8')

In [31]:
df_ords_prods_merge['order_day_of_week'] = df_ords_prods_merge['order_day_of_week'].astype('int8')

In [32]:
df_ords_prods_merge['order_hour_of_day'] = df_ords_prods_merge['order_hour_of_day'].astype('int8')

In [33]:
df_ords_prods_merge['days_since_prior_order'] = df_ords_prods_merge['days_since_prior_order'].astype('float16')

In [34]:
df_ords_prods_merge['product_id'] = df_ords_prods_merge['product_id'].astype('float16')

In [35]:
df_ords_prods_merge['add_to_cart_order'] = df_ords_prods_merge['add_to_cart_order'].astype('int8')

In [36]:
df_ords_prods_merge['reordered'] = df_ords_prods_merge['reordered'].astype('int8')

In [37]:
df_ords_prods_merge['aisle_id'] = df_ords_prods_merge['aisle_id'].astype('int16')

In [38]:
df_ords_prods_merge['department_id'] = df_ords_prods_merge['department_id'].astype('int8')

In [39]:
df_ords_prods_merge['prices'] = df_ords_prods_merge['prices'].astype('float16')

In [40]:
df_ords_prods_merge['max_order'] = df_ords_prods_merge['max_order'].astype('int16')

In [41]:
df_ords_prods_merge['mean_price'] = df_ords_prods_merge['mean_price'].astype('float16')

In [42]:
df_ords_prods_merge['median_days_since_previous_order'] = df_ords_prods_merge['median_days_since_previous_order'].astype('float16')

In [43]:
# Check updated datatypes for df_ords_prods_merge
df_ords_prods_merge.dtypes

order_id                              int64
user_id                               int64
order_number                           int8
order_day_of_week                      int8
order_hour_of_day                      int8
days_since_prior_order              float16
new_customer                           bool
product_id                          float16
add_to_cart_order                      int8
reordered                              int8
product_name                         object
aisle_id                              int16
department_id                          int8
prices                              float16
price_range_loc                      object
busiest_days                         object
busiest_period_of_day                object
max_order                             int16
loyalty_flag                         object
mean_price                          float16
spending_flag                        object
median_days_since_previous_order    float16
Frequency_flag                  

In [48]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True       196.0                  1   
1                    15.0         False       196.0                  1   
2                    21.0         False       196.0                  1   
3                    29.0         False       196.0                  1   
4                    28.0         False       196.0                  1   

   reordered  ... prices    price_range_loc    busiest_days  \
0          0  ...    9.0  Mid-range product  Regularly busy   
1          1  ...    9.0  Mid-range product    Slowest days   
2          1  ...    9.0  Mid-range product    Slowest days   
3          1  ...    9.0  Mid-range product    Slowest days   
4          1  ...    9.0  Mid-range product    Slowest days   

   busiest_period_of_day max_order  loyalty_flag mean_price  spending_flag  \
0         Average orders        10  New customer   6.367188    Low spender   
1         Average orders        10  New customer   6.367188    Low spender   
2            Most orders        10  New customer   6.367188    Low spender   
3         Average orders        10  New customer   6.367188    Low spender   
4            Most orders        10  New customer   6.367188    Low spender   

  median_days_since_previous_order         Frequency_flag  
0                             20.5  Non-frequent customer  
1                             20.5  Non-frequent customer  
2                             20.5  Non-frequent customer  
3                             20.5  Non-frequent customer  
4                             20.5  Non-frequent customer  

[5 rows x 23 columns]

In [49]:
# Merge the 2 dataframes df_ords_prods_merge & df_customers

df_merged = df_ords_prods_merge.merge(df_customers, on = 'user_id')


In [50]:
# Check new merged dataframe
df_merged.shape

(32404859, 32)

In [51]:
df_merged.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True       196.0                  1   
1                    15.0         False       196.0                  1   
2                    21.0         False       196.0                  1   
3                    29.0         False       196.0                  1   
4                    28.0         False       196.0                  1   

   reordered  ...         Frequency_flag  first_name  last_name  gender  \
0          0  ...  Non-frequent customer       Linda     Nguyen  Female   
1          1  ...  Non-frequent customer       Linda     Nguyen  Female   
2          1  ...  Non-frequent customer       Linda     Nguyen  Female   
3          1  ...  Non-frequent customer       Linda     Nguyen  Female   
4          1  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined  no_of_dependants marital_status  income  
0  Alabama  31   2/17/2019                 3        married   40423  
1  Alabama  31   2/17/2019                 3        married   40423  
2  Alabama  31   2/17/2019                 3        married   40423  
3  Alabama  31   2/17/2019                 3        married   40423  
4  Alabama  31   2/17/2019                 3        married   40423  

[5 rows x 32 columns]

# Export merged file as pkl file

In [52]:
df_merged.to_pickle(os.path.join(path,'02 Data', 'Prepared Data', 'df_merged_final.pkl'))